quinn-udp/src/windows.rs

```diff
@@ -111,22 +300,125 @@ impl UdpSocketState {
+                    socket.0.as_raw_socket() as usize,
+                    &mut wsa_msg,
+                    &mut len,
+                    ptr::null_mut(),
+                    None,
+                );
+                if rc == -1 {
+                    return Err(io::Error::last_os_error());
+                }
+            }
+
+            // FIXME add Safety: ?
+            let addr = unsafe {
+                let (_, addr) = socket2::SockAddr::try_init(|addr_storage, len| {
+                    *len = mem::size_of_val(&source) as _;
+                    ptr::copy_nonoverlapping(&source, addr_storage as _, 1);
+                    Ok(())
+                })?;
+                addr.as_socket()
+            };
+
+            // Decode control messages (PKTINFO and ECN)
+            let mut ecn_bits = 0;
+            let mut dst_ip = None;
+
+            let cmsg_iter = unsafe { cmsg::Iter::new(&wsa_msg) };
+            for cmsg in cmsg_iter {
+                match (cmsg.cmsg_level, cmsg.cmsg_type) {
+                    (WinSock::IPPROTO_IP, WinSock::IP_PKTINFO) => {
+                        let pktinfo = unsafe { cmsg::decode::<WinSock::IN_PKTINFO>(cmsg) };
+                        let mut destination: WinSock::SOCKADDR_IN = unsafe { mem::zeroed() };
+                        destination.sin_family = WinSock::AF_INET as _;
+                        destination.sin_addr = pktinfo.ipi_addr;
+                        let (_, address) = unsafe {
+                            socket2::SockAddr::try_init(|addr_storage, len| {
+                                *len = mem::size_of_val(&destination) as _;
+                                std::ptr::copy_nonoverlapping(&destination, addr_storage as _, 1);
+                                Ok(())
+                            })?
+                        };
+                        dst_ip = address.as_socket().map(|s| s.ip());
+                    }
+                    (WinSock::IPPROTO_IPV6, WinSock::IPV6_PKTINFO) => {
+                        let pktinfo = unsafe { cmsg::decode::<WinSock::IN6_PKTINFO>(cmsg) };
+                        let mut destination: WinSock::SOCKADDR_IN6 = unsafe { mem::zeroed() };
+                        destination.sin6_family = WinSock::AF_INET6 as _;
+                        destination.sin6_addr = pktinfo.ipi6_addr;
+                        let (_, address) = unsafe {
+                            socket2::SockAddr::try_init(|addr_storage, len| {
+                                *len = mem::size_of_val(&destination) as _;
+                                std::ptr::copy_nonoverlapping(&destination, addr_storage as _, 1);
+                                Ok(())
+                            })?
+                        };
+                        dst_ip = address.as_socket().map(|s| s.ip());
+                    }
+                    (WinSock::IPPROTO_IP, WinSock::IP_ECN) => {
+                        ecn_bits = unsafe { cmsg::decode::<u8>(cmsg) };
```

Comment:
As above, can we justify this choice of payload type?

Reply:
Nice catch, this is a bug and should be a u32, i will fix the code and also link to a documentation if i found one. Otherwise i can also link to msquic userland implementation.